### ISSUE TITLE
CuPy for HIP vs ClPy for OpenCL for Radeon VII

### ISSUE THREAD

**Issue by avimanyu786** [cupy/cupy#2179]
Hello,

I've been recently trying to run my existing CuPy code (that currently runs on an NVIDIA Titan X (Maxwell) GPU on another new system with an AMD Radeon VII GPU for the same project. On the NVIDIA based system, I was successfully able to install CuPy within a Conda env with Pip.

For the new AMD based system, so far I have followed the instructions as specified at PR #1094 and have been trying to install it. I was able to set it up but was unable to execute `import cupy` successfully. It kept giving an error for hipsparse: `undefined symbol: _ZN8hip_impl8hip_initEv`. At that time I was using ROCm 2.2. After upgrading to ROCm 2.3, I now get a new error for cupy/core/core.so: `undefined symbol: hipModuleGetGlobal` when trying to test `import cupy`. Is there a workaround for this?

Yesterday, I also came to know about [ClPy](https://github.com/fixstars/clpy/). I also found that it received [ROCm OpenCL support](https://github.com/fixstars/clpy/pull/222) very recently. Should I be going for ClPy instead? As of now, is ClPy a better option performance wise? When can we expect _CuPy for HIP_ to be updated like _ClPy_?

Thank you so much for developing CuPy for both platforms! Looking forward to your valuable feedback.

Best,
Avimanyu

**Comment by avimanyu786** [comment]
I upgraded to ROCm 2.4 today and now am not getting any errors for CuPy on HIP when I execute the line `import cupy` :D !!! When I try to display the current device, it shows as `<CUDA Device 0>`. 

```
import cupy
x = cupy.array([1, 2, 3, 4, 5])
print(x)
print(x.device)
exit()
```
The above now gives the following output:

```
[1 2 3 4 5]
<CUDA Device 0>
Process finished with exit code 0
```

Excited to explore more!

**Comment by okuta** [comment]
Thank you for testing CuPy HIP branch #1094. It branch has many issues.
In now, using ClPy is a smart option.


I will continue to support ROCm. However, work is taking time because ROCm is not stable for CuPy.
CuPy uses internal interface of ROCm to support just in time compile.

**Comment by leofang** [comment]
@okuta are you able to comment perhaps briefly on the internal, technical differences between your branch and CIPy? I’m curious. In the future we might have a project running on AMD GPUs, so your comments could be very useful for potential contributors. Thanks.

**Comment by okuta** [comment]
I updated #1094 for ROCm 2.6+.

**Comment by okuta** [comment]
@leofang CuPy uses [ROCm](https://rocm-documentation.readthedocs.io/en/latest/index.html) to support AMD GPU. ClPy uses OpenCL to support AMD GPU and other devices.

The big difference is whether you can use C++ for kernel code.
ClPy applies a patch to convert C++ code to C because OpenCL does not support C++.

If user want to use AMD GPU, using ROCm is better IMO.
However, ClPy is useful when you want to calculate on devices that only support OpenCL.

**Comment by leofang** [comment]
@okuta thank you very much for your enlightening comment! I knew CIPy says it's backed by OpenCL, but I always get confused between that and ROCm (for people growing up in a CUDA world like me, see [this link](https://rocm-documentation.readthedocs.io/en/latest/Programming_Guides/Programming-Guides.html#rocm-languages) for their relation 😄). 

I agree taking the route of ROCm is better, at least in the HPC world where AMD architecture is catching up. This effort will allow us to experiment CuPy on new AMD systems in the future. Your hard work is very much appreciated, @okuta!

**Comment by okuta** [comment]
CuPy [v7.0.0rc1](https://github.com/cupy/cupy/releases/tag/v7.0.0rc1) was released (include #1094).
It supports HIP environment!
